### PR TITLE
Fix/skip tests for mxnet backend

### PR DESCRIPTION
**Pull request by sandeep-krishnamurthy**
First step towards running all keras tests mxnet backend.
Fixed or skipped tests cases in following modules. All missing functionalities are captured in this issue - https://github.com/deep-learning-tools/keras/issues/20
* tests/integration_tests/
* tests/keras/applications/
* tests/keras/backend/
* tests/keras/datasets/
* tests/keras/engine/
* tests/keras/legacy/
* tests/keras/preprocessing
* tests/keras/utils/

With this PR, we will enable around 200 keras test cases for mxnet backend.

@jesterhazy

### CODE REVIEW

File: keras/backend/mxnet_backend.py [hunk 1]
```diff
@@ -2581,22 +2584,31 @@ def in_train_phase(x, alt, training=None):
...
         the `training` flag defaults to `K.learning_phase()`.
     """
+    uses_learning_phase = False
+
     if training is None:
         training = learning_phase()
+        uses_learning_phase = True
 
-    if training is 1:
+    if training is 1 or training is True:
```
Comment (jesterhazy) [inline keras/backend/mxnet_backend.py]: `if training:`

File: keras/backend/mxnet_backend.py [hunk 2]
```diff
@@ -2581,22 +2584,31 @@ def in_train_phase(x, alt, training=None):
...
         if callable(x):
-            return x()
+            res = x()
         else:
-            return x
-    elif training is 0:
+            res = x
+        if isinstance(x, KerasSymbol):
+            uses_learning_phase = True
+    elif training is 0 or training is False:
```
Comment (jesterhazy) [inline keras/backend/mxnet_backend.py]: `else:`

File: keras/backend/mxnet_backend.py [hunk 3]
```diff
@@ -2581,22 +2584,31 @@ def in_train_phase(x, alt, training=None):
...
         if callable(alt):
-            return alt()
+            res = alt()
         else:
-            return alt
+            res = alt
+    else:
```
Comment (jesterhazy) [inline keras/backend/mxnet_backend.py]: is this a real case? i think we only get here if training is int > 1

File: tests/integration_tests/test_temporal_data_tasks.py [hunk 4]
```diff
@@ -11,6 +11,7 @@
 import keras
 
 
+@pytest.mark.skipif(K.backend() == 'mxnet', reason='MXNet backend do not support GRU')
```
Comment (jesterhazy) [inline tests/integration_tests/test_temporal_data_tasks.py]: does not

File: tests/integration_tests/test_temporal_data_tasks.py [hunk 5]
```diff
@@ -44,6 +45,7 @@ def test_temporal_classification():
     model = Sequential.from_config(config)
 
 
+@pytest.mark.skipif(K.backend() == 'mxnet', reason='MXNet backend do not support RNN')
```
Comment (jesterhazy) [inline tests/integration_tests/test_temporal_data_tasks.py]: does not

File: tests/integration_tests/test_temporal_data_tasks.py [hunk 6]
```diff
@@ -74,6 +76,7 @@ def test_temporal_classification_functional():
     assert(history.history['acc'][-1] >= 0.8)
 
 
+@pytest.mark.skipif(K.backend() == 'mxnet', reason='MXNet backend do not support LSTM')
```
Comment (jesterhazy) [inline tests/integration_tests/test_temporal_data_tasks.py]: does not

File: tests/integration_tests/test_temporal_data_tasks.py [hunk 7]
```diff
@@ -119,6 +122,7 @@ def test_3d_to_3d():
     assert(history.history['loss'][-1] < 1.)
 
 
+@pytest.mark.skipif(K.backend() == 'mxnet', reason='MXNet backend do not support LSTM')
 @keras_test
 def test_stacked_lstm_char_prediction():
     '''
```
Comment (jesterhazy) [inline tests/integration_tests/test_temporal_data_tasks.py]: does not

File: tests/integration_tests/test_temporal_data_tasks.py [hunk 8]
```diff
@@ -168,6 +172,7 @@ def test_stacked_lstm_char_prediction():
     assert(generated == alphabet)
 
 
+@pytest.mark.skipif(K.backend() == 'mxnet', reason='MXNet backend do not support embedding layers')
```
Comment (jesterhazy) [inline tests/integration_tests/test_temporal_data_tasks.py]: does not

File: tests/integration_tests/test_vector_data_tasks.py [hunk 9]
```diff
@@ -43,6 +44,7 @@ def test_vector_classification():
     model = Sequential.from_config(config)
 
 
+@pytest.mark.skipif(K.backend() == 'mxnet', reason='MXNet backend do not support Sparse')
```
Comment (jesterhazy) [inline tests/integration_tests/test_vector_data_tasks.py]: does not

File: tests/keras/engine/test_training.py [hunk 10]
```diff
@@ -779,8 +786,8 @@ def test_model_with_partial_loss():
 
 
 @keras_test
-@pytest.mark.skipif((K.backend() == 'cntk'),
-                    reason='cntk does not support external loss yet')
+@pytest.mark.skipif((K.backend() == 'cntk' or K.backend() == 'mxnet'),
+                    reason='cntk/mxnet does not support external loss yet')
```
Comment (jesterhazy) [inline tests/keras/engine/test_training.py]: do not

File: tests/keras/utils/vis_utils_test.py [hunk 11]
```diff
@@ -9,8 +9,11 @@
 from keras.layers import TimeDistributed
 from keras.models import Sequential
 from keras.utils import vis_utils
+from keras import backend as K
 
 
+@pytest.mark.skipif(K.backend() == 'mxnet',
+                    reason='MXNet backend do not LSTM yet')
```
Comment (jesterhazy) [inline tests/keras/utils/vis_utils_test.py]: does not

File: keras/backend/mxnet_backend.py (hunk 2 above)
Comment (sandeep-krishnamurthy) [reply keras/backend/mxnet_backend.py]: We can have the training parameter to be "0 or 1 or False or True or Placeholder". Hence, if, elif, else.
Simplified it more.

File: keras/backend/mxnet_backend.py (hunk 3 above)
Comment (sandeep-krishnamurthy) [reply keras/backend/mxnet_backend.py]: training can be a Keras Placeholder.

File: tests/integration_tests/test_temporal_data_tasks.py (hunk 4 above)
Comment (sandeep-krishnamurthy) [reply tests/integration_tests/test_temporal_data_tasks.py]: Fixed. Also in all other places.

File: tests/keras/engine/test_training.py (hunk 10 above)
Comment (sandeep-krishnamurthy) [reply tests/keras/engine/test_training.py]: done.

**Comment by sandeep-krishnamurthy** [comment]
@yuruofeifei @jesterhazy 
Added 3 more commits:
1. Fixing bugs in temporal_padding, spatial_2dpadding and spatial_3dpadding.
2. Fixing bugs in batchnorm - added momentum parameter.
3. Enable more test cases (around 100 more tests).
